fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -164,6 +177,232 @@ private boolean isMVApplicable(OptExpression mvExpression,
+                return false;
+            }
+            Table leftTable = usedColumnsToTable.get(leftColumnRefSet.get());
+            if (leftTable.getColumns().stream().allMatch(column -> column.isAllowNull())) {
+                // should have at least one nullable column
+                return  false;
+            }
+            List<ColumnRefOperator> leftJoinColumnRefs =
+                    leftColumnRefSet.get().getColumnRefOperators(materializationContext.getQueryRefFactory());
+            joinColumnRefs.set(0, leftJoinColumnRefs);
+
+            Optional<ColumnRefSet> rightColumnRefSet = usedColumnsToTable.keySet()
+                    .stream().filter(columnSet -> rightColumns.containsAll(columnSet)).findFirst();
+            if (!rightColumnRefSet.isPresent()) {
+                return false;
+            }
+            Table rightTable = usedColumnsToTable.get(rightColumnRefSet.get());
+            if (rightTable.getColumns().stream().allMatch(column -> column.isAllowNull())) {
+                // should have at least one nullable column
+                return  false;
+            }
+            List<ColumnRefOperator> rightJoinColumnRefs =
+                    rightColumnRefSet.get().getColumnRefOperators(materializationContext.getQueryRefFactory());
+            joinColumnRefs.set(1, rightJoinColumnRefs);
+        }
+        return isCompatible;
+    }
+
+    private boolean isUniqueColumns(Table table, List<String> columnNames) {
+        if (table.isOlapTable()) {
+            OlapTable olapTable = (OlapTable) table;
+            if (olapTable.getKeysType() == KeysType.PRIMARY_KEYS || olapTable.getKeysType() == KeysType.UNIQUE_KEYS) {
+                List<String> keyColumnNames =
+                        olapTable.getKeyColumns().stream().map(column -> column.getName()).collect(Collectors.toList());
+                return columnNames.containsAll(keyColumnNames);
+            }
+        }
+        return table.hasUniqueConstraints() && table.getUniqueConstraints().stream().anyMatch(
+                uniqueConstraint -> columnNames.containsAll(uniqueConstraint.getUniqueColumns()));
+    }
+
+    private boolean isSupportedPredicate(
+            ScalarOperator onPredicate, ColumnRefFactory columnRefFactory, Map<Integer, ColumnRefSet> joinColumns) {
+        List<ScalarOperator> conjuncts = Utils.extractConjuncts(onPredicate);
+        List<ScalarOperator> binaryPredicates = conjuncts.stream()
+                .filter(conjunct -> Utils.isColumnEqualBinaryPredicate(conjunct)).collect(Collectors.toList());
+        if (binaryPredicates.isEmpty()) {
+            return false;
+        }
+
+        ColumnRefSet usedColumns = Utils.compoundAnd(binaryPredicates).getUsedColumns();
+        for (int columnId : usedColumns.getColumnIds()) {
+            int relationId = columnRefFactory.getRelationId(columnId);
+            if (relationId == -1) {
+                // not use the column from table scan, unsupported
```

Comment:
in the future we could turn these comment into a "trace log", to help the diagnosis of rewrite ?

Reply:
yes